### ISSUE TITLE
Run Dino in background

### ISSUE THREAD

**Issue by schiessle** [dino/dino#299]
Dino is a really nice looking and modern XMPP client. What stops me from using it on a daily basis is that I can't run it in the background. I would like to autostart it during login in the background so that I get notification as soon as someone writes me a message or mention me in a group chat.

Is something like this already planed?

**Comment by d9h20f** [comment]
I think any operating system and DE allows for autostarting applications. You can move it to another workspace (or just minimise it) if you don't want to see the window.

**Comment by GladOSkar** [comment]
I think this is a duplicate of #98

**Comment by schiessle** [comment]
You could also look how Polari, the Gnome IRC client, does it the "Gnome way" without the need of a system tray.

**Comment by benrob0329** [comment]
@schiessle i.e. just start minimized?

**Comment by schiessle** [comment]
@benrob0329 I would call it "starting in the background". In Gnome (what I care most) it shouldn't be visible in the activities overview, still notifications should pop up. If I close the chat window it should again go back in the background. I think Polari handles it quite well.

In a non Gnome setup, background could be "minimize to tray". If Dino wants to target non Gnome desktops.

**Comment by gllmhyt** [comment]
> I think Polari handles it quite well and I think it could be worth looking at how they did it.

[Polari](https://wiki.gnome.org/Apps/Polari) has [this file](https://gitlab.gnome.org/GNOME/polari/blob/master/data/org.gnome.Polari.Autostart.desktop) that starts the client (`gapplication action org.gnome.Polari start-client`) hidden (`NoDisplay=True`).

```
[Desktop Entry]
Type=Application
Exec=gapplication action org.gnome.Polari start-client
NoDisplay=True
```

Moreover, Polari has the ability to be ran in background then avoid annoying leave/join and so on. [Here are the commits](: https://bugzilla.gnome.org/show_bug.cgi?id=770750#c14) that made it possible.

> When set up to run in the background, the application:
> * sets itself up to start in the background on login
> * keeps running after the last window has been closed
> * can be stopped via the 'quit' action in the application menu
> ([source](https://gitlab.gnome.org/GNOME/polari/commit/00ab02c))

**Comment by notramo** [comment]
@fiaxh any progress on this?
I suggest that until systray support will be added, it could run in the background simply without system tray:
- The process is autostarted in background, and if a **message is received**, a **notification** would appear.
- Optionally, the **window is raised** and (optionally) **focused on new message**.
- If **there is a running `dino` process**, and the user starts it with menu icon, the *new process* would only **instruct the current** to **raise the window**.
- A hotkey would raise the window (if there isn't global hotkey support, then use the keybinding of DE/WM with a command: `dino --raise`)

This would be useful, because I install Linux for people who aren't experts nor power users. I want them to be always online when their computer is on. But they don't want their taskbar to be cluttered, so they would simply close the window when autostarted.

**Comment by trymeouteh** [comment]
+1 for system tray support and the ability to minimise or close Dino to the system tray.

**Comment by notramo** [comment]
You can try out Matrix, an open-source, E2EE alternative to XMPP, until it is fixed in Dino. (Matrix also has cross-device signing, which makes verifying devices much easier.)
There are many Matrix clients, which support this feature, has modern UI, and much more.
matrix.org/clients

**Comment by crabvk** [comment]
@notramo Don't spam here, sir.

**Comment by Blueshawk** [comment]
Minimize to tray, installing with an autostart for the most used distros, daemonizing and system notification integrations are basics these days. I applaud this work, but feel that this is keeping dino marginalized in many distros. 

Currently using mint 21 xfce

**Comment by methuselahin** [comment]
> Minimize to tray, installing with an autostart for the most used distros, daemonizing and system notification integrations are basics these days. I applaud this work, but feel that this is keeping dino marginalized in many distros.
> 
> Currently using mint 21 xfce

yes this is very crucial option. I mean this is just basic functionality.

**Comment by TriggerDingus** [comment]
I just installed Dino as a replacement for a different client and was looking for system tray and autostart options and couldn't find anything and I thought I was just missing something. From the age of this thread, it suggests to me this isn't in the design roadmap. Strange, for a chat client..

**Comment by licaon-kter** [comment]
@TriggerDingus: start Dino (from your wm/de autostart settings) as:
`dino --gapplication-service`

So you can close and it stays running, no tray icon, but notifications, etc

**Comment by Buntbart** [comment]
Unfortunately this is not equivalent to a tray icon, because in your variant there is no indication of newly received messages in case you missed the notification. A tray icon would show that and would allow to open dino directly.

**Comment by licaon-kter** [comment]
@Buntbart I didn't say it was

@TriggerDingus if you test from the command line be sure to add `&` at the end so it stays in memory

Notifications should be shown by your notification thing, eg. Dunst or whatever you have that's standards compliant.

To open Dino you start Dino normally (that command is only for the first start)

**Comment by iclanzan** [comment]
I just tried this `--gapplication-service` approach on Pinephone with Dino 0.4.0 and it seems to work so far!

**Comment by Jieiku** [comment]
This is a hard requirement for me.

I can solve the issue of the tray icon on linux by using kdocker, so all that is really missing is an option in Dino to minimize instead of close.

**Comment by hyphenrf** [comment]
is there an update on this with the merging of background apps feature in gnome 46?

**Comment by rmader** [comment]
Looks like there are two independent PRs for this now:
 - https://github.com/dino/dino/pull/1826
 - https://github.com/dino/dino/pull/1828

**Comment by kousu** [comment]
> Looks like there are two independent PRs for this now:
> 
>     * [RFC: Add autostart and background support via libportal #1826](https://github.com/dino/dino/pull/1826)
> 
>     * [Systray Support (Linux) #1828](https://github.com/dino/dino/pull/1828)

Oop, yes so there are. Wow we submitted right past each other, huh. I missed the other PR because I was looking back at #98 and #98 emphasized that Dino was not just for Gnome. I didn't know about Gnome's new background apps tray, that's nice. In principle there is no problem swapping the code from #1826 to the Gnome case in #1828 .

**Comment by Jieiku** [comment]
Try DinoX https://github.com/rallep71/dinox

**Comment by hyphenrf** [comment]
I don't know about you but I wouldn't trust a heap of unrestrained llm slop to run my secure messaging client

**Comment by Jehan** [comment]
Also advertising another project (very likely a fork, so most of their "work" is based on the original) inside a project's report like this is a very lousy move.

FLOSS projects are working together and it's fine using several software. We don't need playing the competitive game or trying to divert happy "users" of other software.

**Comment by Jieiku** [comment]
Thanks @hyphenrf for mentioning the AI generated code, I was not aware, also @jehan I only just recently found DinoX and have been testing it out the last few days. Not trying to divert happy users either, I simply said to try DinoX because it has the feature being asked for in this issue report that dates back to 2018, maybe it can fill a gap for some people that would otherwise not be using Dino at all! I am such a user and have been sticking with Pidgin with the lurch plugin for a LOT of years because of system tray support, so having DinoX has been nice compared to using a very outdated Pidgin.

**Comment by kousu** [comment]
> I don't know about you but I wouldn't trust a heap of unrestrained llm slop to run my secure messaging client

strong agree. but would you trust it to *find* security problems? because most of these seem like genuine issues in Dino and I would like to know the client I am using has them fixed:

- https://github.com/rallep71/dinox/blob/23f7d59a92f5d8d68121b46361e2f6de95701215/SECURITY_AUDIT.md
- https://github.com/rallep71/dinox/blob/8921eef1f414c916d7e8f2a39224c888ab77851c/SECURITY_AUDIT.md
- https://github.com/rallep71/dinox/blob/0127064d499a35d9b204a57c4169c4810a552fd2/SECURITY_AUDIT.md

> plugins/omemo/src/native/helper.c:159 	Heap corruption: free(md) called on gcry_md_read() internal pointer. gcry_md_read() returns a pointer to an internal buffer that must not be freed. Also in upstream Dino.

**Comment by hyphenrf** [comment]
I'm not arguing human written software is perfectly secure, if there is any implication of that. This output you quote is no different than running a fuzzer on dino, except way more resource hungry.

What I am saying is, there's a thread of accountability and explainability which gets broken when the code you're dealing with is basically LLM artifact dumps. It introduces ambiguity and blurs ownership. Even codegen tools would be replayable and debuggable. The process is at human speed and scale, social dynamics are present, so malicious actors lose reputation.. How do you reconcile this whole infrastructure with slop coding? Except by reckless abandon... In any case this issue is probably not the place to discuss that and we've gone way offtopic I reckon.

**Comment by kousu** [comment]
No, of course not. I'm not saying you implied that! But I am saying don't throw the baby out with the bath water. 
Could semgrep have caught that one? Could any tool? Maybe? I've posted https://github.com/dino/dino/issues/1859 to follow up; it seems like semgrep might actually not be viable because no one has ever (or yet) taught it about vala.

> In any case this issue is probably not the place to discuss that and we've gone way offtopic I reckon.

Yeah, you're right, of course. 🦫

**Comment by fiaxh** [comment]
> but would you trust it to find security problems? because most of these seem like genuine issues in Dino

None of the critical / high severity problems are actually valid or relevant in practice. More details are in https://github.com/dino/dino/issues/1859#issuecomment-4225885125, but since this is a sensitive topic, I didn't want to leave it uncommented here.
